transport/graphsync/graphsync.go

```diff
@@ -164,6 +166,69 @@ func (t *Transport) OpenChannel(ctx context.Context,
 	return nil
 }
 
+// Get the extension data for sending a Restart message, depending on the
+// protocol version of the peer
+func (t *Transport) getRestartExtension(ctx context.Context, p peer.ID, channel datatransfer.ChannelState) ([]graphsync.ExtensionData, error) {
+	if channel == nil {
+		return nil, nil
+	}
+
+	// Get the peer's protocol version
+	protocol, err := t.peerProtocol.Protocol(ctx, p)
```

Comment:
So I want to point out something that's a potential reason to augment the Graphsync extension as well: in the case of a pull request, you may still not have this cached even for a restart, as the only communication may be over graphsync. I don't know whether this matters a lot, but the idea would be to maintain your protocol list outside the network implementation, so you can update it for an incoming GS request/response as well as well. The only penalty is a new libp2p message I guess, so not the worst thing but still.

Reply:
I see what you mean, that may save us an extra round trip.

If I understand correctly, if we add another extension protocol, then whenever we send a message, we also include an encoded payload for that extension. Is that correct? So for example if we have an extension for go-data-transfer-v1.2, the message will look like:
```
<message data>
extensions:
  fil/data-transfer/1.2: <encoded data>
  fil/data-transfer/1.1: <encoded data>
  fil/data-transfer:     <encoded data>
```

Maybe easier to explain with a screenshot from the debugger:

<img width="1095" alt="Screen Shot 2021-10-01 at 6 46 24 PM" src="https://user-images.githubusercontent.com/169124/135657685-93a64957-dfaf-4312-81a4-e0d74ec1ff21.png">
:

Reply:
yes it's an additional copy of the exact same mesage... so arguably this ones a bit a bummer, especially since there's no difference in the message format.... it's literally just the extension name.

BTW we can probably deprecate 1_0 at this point -- hasn't existed since pre-mainnet.

Reply:
Also Aarsh pointed out that for a pull request it's the data receiver who *sends* the extension, so the data receiver is not going to know the protocol of the data sender.
So I think it makes most sense not to use an additional extension for now.